controllers/base.go

```diff
@@ -174,10 +165,122 @@ func (bmrc *baseMachineReconcileContext) releaseHardware() error {
+	powerOffAction := rufiov1.HardPowerOff
+	controller := true
+	bmcJob := &rufiov1.BMCJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-poweroff", bmrc.tinkerbellMachine.Name),
+			Namespace: bmrc.tinkerbellMachine.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+					Kind:       "TinkerbellMachine",
+					Name:       bmrc.tinkerbellMachine.Name,
+					UID:        bmrc.tinkerbellMachine.ObjectMeta.UID,
+					Controller: &controller,
+				},
+			},
+		},
+		Spec: rufiov1.BMCJobSpec{
+			BaseboardManagementRef: rufiov1.BaseboardManagementRef{
+				Name:      hardware.Spec.BMCRef.Name,
+				Namespace: bmrc.tinkerbellMachine.Namespace,
+			},
+			Tasks: []rufiov1.Task{
+				{
+					PowerAction: &powerOffAction,
+				},
+			},
+		},
+	}
+
+	if err := bmrc.client.Create(bmrc.ctx, bmcJob); err != nil {
+		return fmt.Errorf("creating BMCJob: %w", err)
+	}
+
+	return nil
+}
+
+// getBMCJob fetches the BMCJob with name JName.
+func (bmrc *baseMachineReconcileContext) getBMCJob(jName string, bmj *rufiov1.BMCJob) error {
+	namespacedName := types.NamespacedName{
+		Name:      jName,
+		Namespace: bmrc.tinkerbellMachine.Namespace,
+	}
+
+	if err := bmrc.client.Get(bmrc.ctx, namespacedName, bmj); err != nil {
+		return fmt.Errorf("GET BMCJob: %w", err)
+	}
+
+	return nil
+}
+
 // DeleteMachineWithDependencies removes template and workflow objects associated with given machine.
 func (bmrc *baseMachineReconcileContext) DeleteMachineWithDependencies() error {
 	bmrc.log.Info("Removing machine", "hardwareName", bmrc.tinkerbellMachine.Spec.HardwareName)
+	// Fetch hardware for the machine.
+	hardware := &tinkv1.Hardware{}
+	if err := bmrc.getHardwareForMachine(hardware); err != nil {
+		return err
+	}
+
+	if err := bmrc.removeDependencies(hardware); err != nil {
+		return err
+	}
+
+	// The hardware BMCRef is nil.
+	// Remove finalizers and let machine object delete.
+	if hardware.Spec.BMCRef == nil {
+		bmrc.log.Info("Hardware BMC reference not present; skipping hardware power off",
+			"BMCRef", hardware.Spec.BMCRef, "Hardware", hardware.Name)
+
+		return bmrc.removeFinalizer()
+	}
+
+	// Fetch a poweroff BMCJob for the machine.
+	// If Job not found, we remove dependencies and create job.
+	bmcJob := &rufiov1.BMCJob{}
+	jobName := fmt.Sprintf("%s-poweroff", bmrc.tinkerbellMachine.Name)
+
+	err := bmrc.getBMCJob(jobName, bmcJob)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return bmrc.createPowerOffJob(hardware)
+		}
+
+		return fmt.Errorf("get bmc job for machine: %w", err)
+	}
+
+	// Check the Job conditions to ensure the power off job is complete.
+	if bmcJob.HasCondition(rufiov1.JobCompleted, rufiov1.ConditionTrue) {
+		return bmrc.removeFinalizer()
+	}
 
+	if bmcJob.HasCondition(rufiov1.JobFailed, rufiov1.ConditionTrue) {
+		return fmt.Errorf("bmc job %s/%s failed", bmcJob.Namespace, bmcJob.Name) // nolint:goerr113
+	}
```

Comment:
this feels like something that could benefit from being abstracted to its own method. this could improve maintainability; make it easier to remove, modify, etc.